File: frameworks/go/echo/Implementation/query.go
package implementation

import (
	"net/http"

	"github.com/labstack/echo/v5"
)

type QueryOne struct {
	Page int `query:"page" json:"page"`
}

// Search is query.many's eight values, which forms.urlencoded posts as a form.
type Search struct {
	Page     int    `query:"page" form:"page" json:"page"`
	Size     int    `query:"size" form:"size" json:"size"`
	Status   string `query:"status" form:"status" json:"status"`
	Category string `query:"category" form:"category" json:"category"`
	Sort     string `query:"sort" form:"sort" json:"sort"`
	Q        string `query:"q" form:"q" json:"q"`
	MinPrice int    `query:"minPrice" form:"minPrice" json:"minPrice"`
	MaxPrice int    `query:"maxPrice" form:"maxPrice" json:"maxPrice"`
}

// queryRoutes bind the query string with echo.BindQueryParams, where a field's query tag names
// the parameter and its type is the conversion.
func queryRoutes(e *echo.Echo, p *Payloads) {
	e.GET("/query/one", func(c *echo.Context) error {
		var query QueryOne
		if err := echo.BindQueryParams(c, &query); err != nil {
			return err
		}
		return c.JSON(http.StatusOK, Echoed[QueryOne]{&p.Small, query})
	})

	e.GET("/query/many", func(c *echo.Context) error {
		var query Search
		if err := echo.BindQueryParams(c, &query); err != nil {
			return err
		}
		return c.JSON(http.StatusOK, Echoed[Search]{&p.Small, query})
	})
}

File: frameworks/python/django-asgi/Implementation/views/query.py
from django import forms
from django.conf import settings
from django.http import JsonResponse
from django.views.decorators.http import require_GET

from answers import echoed

# query: the query string bound by a Form, which is Django's binder as well as its validator. Each
# field converts its value, so cleaned_data is already the echo, and a value a field refuses is the
# form's own 400.
P = settings.PAYLOADS


class Page(forms.Form):
    page = forms.IntegerField()


class Search(forms.Form):
    """query.many's eight values, which forms.urlencoded posts as a form."""

    page = forms.IntegerField()
    size = forms.IntegerField()
    status = forms.CharField()
    category = forms.CharField()
    sort = forms.CharField()
    q = forms.CharField()
    minPrice = forms.IntegerField()
    maxPrice = forms.IntegerField()


# rb:handler query.one
@require_GET
async def one(request):
    page = Page(request.GET)
    if not page.is_valid():
        return JsonResponse(page.errors, status=400)
    return JsonResponse(echoed(P.small, page.cleaned_data))


# rb:handler query.many
@require_GET
async def many(request):
    search = Search(request.GET)
    if not search.is_valid():
        return JsonResponse(search.errors, status=400)
    return JsonResponse(echoed(P.small, search.cleaned_data))
